### ISSUE TITLE
Console progress DivideByZero exception

### ISSUE THREAD

**Issue by bioleyl** [Ymagis/ClairMeta#173]
Hello,

I often get an error like that:

```
PKL assets hash check. in PKL_7cb7d1ba-007e-4d31-8fcc-5ba6b214a62b.xml -> 8f2eeddd-ac5b-483d-ba52-4b4a6bb3a353_video.mxf: Check unknown error Traceback (most recent call last): File "C:\worker\Python38\lib\site-packages\clairmeta\dcp_check_base.py", line 152, in run_check check_res = check(*args) File "C:\worker\Python38\lib\site-packages\clairmeta\dcp_check_pkl.py", line 111, in check_assets_pkl_hash self.hash_map[asset_id] = shaone_b64(path, self.hash_callback) File "C:\worker\Python38\lib\site-packages\clairmeta\utils\file.py", line 187, in shaone_b64 callback(file_path, run_size, file_size, time.time() - start) File "C:\worker\Python38\lib\site-packages\clairmeta\utils\file.py", line 124, in __call__ eta_sec = (self._total_size - processed) / (processed / elapsed) ZeroDivisionError: float division by zero
```

Do you have any idea what it can be? Is it because of a network lag (I check DCP over network)?

I am using the last release of ClairMeta

Best regards

Léal

**Comment by remia** [comment]
Hey @bioleyl,

Thanks for the report, looks like we have a bug in the console progression code.

I'm not sure yet, but it looks like the callback get a reported elapsed time of zero that triggers an divide_by_zero exception.

You may want to disable console progress report removing the `-progress` flag while we fix that bug.

**Comment by bioleyl** [comment]
Ok I will remove temporarily the -progress argument.

I have found a DCP that generate all the time that error (It is """small""" ~14GB). Are you interested in it to reproduce?

**Comment by remia** [comment]
Yes that could be useful, is the error happening on a specific assets hash check ?

If that's the case, the XML files + that asset should be enough to reproduce.

**Comment by bioleyl** [comment]
Ok can you give me a "private" way to send a download link to you? There are rights on that DCP and I don't want to send a download link publicly.

**Comment by remia** [comment]
Sure, you can use remiachard - at - gmail.com

**Comment by bioleyl** [comment]
All the assets of the pkl are in error. So I upload everything. You will get a download link by mail in about 15min.

**Comment by remia** [comment]
Thanks @bioleyl, I checked the DCP and didn't get any error.

I'm assuming it's related to the network condition or some other IO factor as you mentioned. I put up #174 with some minor fix that might help, and some additional debug print if that still happen. Can you give it a try and report the error message ?

Thanks

**Comment by bioleyl** [comment]
It seems better but not totally fine. Just to be clear, problems occures only with the -progress argument.

For the good point:
I can see now the progression of the hash check. Before, it failed just before this step.

For the bad point:
After the hash check, it takes a really long time before ending the process (some minutes) and I still got errors:

```
Error(s):
  + PKL_7cb7d1ba-007e-4d31-8fcc-5ba6b214a62b.xml (CAVALES_SHR_S_FR-XX_AL_51_2K_NULL_20210123_NUL_SMPTE_OV)
    . PKL XML syntax and structure check.
      Schema validation error : Element '{http://www.w3.org/2001/XMLSchema}import': Failed to parse the XML resource 'http://www.w3.org/TR/2002/REC-xmldsig-core-20020212/xmldsig-core-schema.xsd'., line 4
      Using schema : http://www.smpte-ra.org/schemas/429-8/2007/PKL
    + 2a9d8f9e-8799-4e5e-a6c2-9c38967bb0c5_sub.mxf
      . PKL assets hash check.
        Check unknown error
        Traceback (most recent call last):
          File "C:\worker\test lib\ClairMeta\clairmeta\dcp_check_base.py", line 152, in run_check
            check_res = check(*args)
          File "C:\worker\test lib\ClairMeta\clairmeta\dcp_check_pkl.py", line 111, in check_assets_pkl_hash
            self.hash_map[asset_id] = shaone_b64(path, self.hash_callback)
          File "C:\worker\test lib\ClairMeta\clairmeta\utils\file.py", line 193, in shaone_b64
            callback(file_path, run_size, file_size, time_cb - start)
          File "C:\worker\test lib\ClairMeta\clairmeta\utils\file.py", line 143, in __call__
            human_size(file_size), file_elapsed, (file_size / 1e6) / file_elapsed)
        ZeroDivisionError: float division by zero

    + CPL_29bea761-1123-49e5-90ae-786c61e0d8cd.xml (CAVALES_SHR_S_FR-EN_AL_51_2K_NULL_20210123_NUL_SMPTE_OV)
      . PKL assets hash check.
        Check unknown error
        Traceback (most recent call last):
          File "C:\worker\test lib\ClairMeta\clairmeta\dcp_check_base.py", line 152, in run_check
            check_res = check(*args)
          File "C:\worker\test lib\ClairMeta\clairmeta\dcp_check_pkl.py", line 111, in check_assets_pkl_hash
            self.hash_map[asset_id] = shaone_b64(path, self.hash_callback)
          File "C:\worker\test lib\ClairMeta\clairmeta\utils\file.py", line 193, in shaone_b64
            callback(file_path, run_size, file_size, time_cb - start)
          File "C:\worker\test lib\ClairMeta\clairmeta\utils\file.py", line 143, in __call__
            human_size(file_size), file_elapsed, (file_size / 1e6) / file_elapsed)
        ZeroDivisionError: float division by zero

    + CPL_a60452a4-14e0-479d-b9f9-4f69e1f6fde7.xml (CAVALES_SHR_S_FR-XX_AL_51_2K_NULL_20210123_NUL_SMPTE_OV)
      . PKL assets hash check.
        Check unknown error
        Traceback (most recent call last):
          File "C:\worker\test lib\ClairMeta\clairmeta\dcp_check_base.py", line 152, in run_check
            check_res = check(*args)
          File "C:\worker\test lib\ClairMeta\clairmeta\dcp_check_pkl.py", line 111, in check_assets_pkl_hash
            self.hash_map[asset_id] = shaone_b64(path, self.hash_callback)
          File "C:\worker\test lib\ClairMeta\clairmeta\utils\file.py", line 193, in shaone_b64
            callback(file_path, run_size, file_size, time_cb - start)
          File "C:\worker\test lib\ClairMeta\clairmeta\utils\file.py", line 143, in __call__
            human_size(file_size), file_elapsed, (file_size / 1e6) / file_elapsed)
        ZeroDivisionError: float division by zero

  + CPL_29bea761-1123-49e5-90ae-786c61e0d8cd.xml (CAVALES_SHR_S_FR-EN_AL_51_2K_NULL_20210123_NUL_SMPTE_OV)
    . CPL XML syntax and structure check.
      Schema validation error : Element '{http://www.w3.org/2001/XMLSchema}import': Failed to parse the XML resource 'http://www.w3.org/TR/2002/REC-xmldsig-core-20020212/xmldsig-core-schema.xsd'., line 4
      Using schema : http://www.smpte-ra.org/schemas/429-7/2006/CPL
    + 2a9d8f9e-8799-4e5e-a6c2-9c38967bb0c5_sub.mxf
      . Subtitle font file exists.
        Subtitle missing font file : ad48b4af-9e02-4cb6-8867-b81ea5e1816d
  + CPL_a60452a4-14e0-479d-b9f9-4f69e1f6fde7.xml (CAVALES_SHR_S_FR-XX_AL_51_2K_NULL_20210123_NUL_SMPTE_OV)
    . CPL XML syntax and structure check.
      Schema validation error : Element '{http://www.w3.org/2001/XMLSchema}import': Failed to parse the XML resource 'http://www.w3.org/TR/2002/REC-xmldsig-core-20020212/xmldsig-core-schema.xsd'., line 4
      Using schema : http://www.smpte-ra.org/schemas/429-7/2006/CPL
Warning(s):
  + CPL_29bea761-1123-49e5-90ae-786c61e0d8cd.xml (CAVALES_SHR_S_FR-EN_AL_51_2K_NULL_20210123_NUL_SMPTE_OV)
    . CPL ContentTitleText and PKL AnnotationText shall match.
      Multi CPLs package shall use a common denominator of all CPL titles as the PKL AnnotationText : CAVALES_SHR_S_FR-EN_AL_51_2K_NULL_20210123_NUL_SMPTE_OV / CAVALES_SHR_S_FR-XX_AL_51_2K_NULL_20210123_NUL_SMPTE_OV

DCP - \\hnas2\NUMERIQUE_FILM\NUMERIQUE_FILM\NON_RESTAURATION_BY_PASS\OFC_controler\Cavales\DCP\CAVALES_SHR_S_FR-XX-EN_AL_51_2K_NULL_20210123_NUL_SMPTE_OV - Check failed
```

**Comment by bioleyl** [comment]
The problem is really around that -progress. If I remove it, everything works fine.

It is not because of the network. I tried it locally and it is the same.

**Comment by bioleyl** [comment]
One other thing. The issue seems to be on windows. I run the command from linux and everything works with the -progress

**Comment by remia** [comment]
Hey @bioleyl,

Looks like you get an error slightly further in the hash check process, just on the last call to the progress report where it should print the completed bar. I updated the PR in case you try again, the print should tell me what is going wrong.

> I run the command from linux and everything works with the -progress

I tested on macOS and got no errors as well, that's good info. I'll need to find a window box to check that.

Last note, these errors are kind of worying, I didn;t get any on macOS so I wonder where they are coming from.

Might be another Windows issue.

```
  + CPL_29bea761-1123-49e5-90ae-786c61e0d8cd.xml (CAVALES_SHR_S_FR-EN_AL_51_2K_NULL_20210123_NUL_SMPTE_OV)
    . CPL XML syntax and structure check.
      Schema validation error : Element '{http://www.w3.org/2001/XMLSchema}import': Failed to parse the XML resource 'http://www.w3.org/TR/2002/REC-xmldsig-core-20020212/xmldsig-core-schema.xsd'., line 4
      Using schema : http://www.smpte-ra.org/schemas/429-7/2006/CPL
    + 2a9d8f9e-8799-4e5e-a6c2-9c38967bb0c5_sub.mxf
      . Subtitle font file exists.
        Subtitle missing font file : ad48b4af-9e02-4cb6-8867-b81ea5e1816d
  + CPL_a60452a4-14e0-479d-b9f9-4f69e1f6fde7.xml (CAVALES_SHR_S_FR-XX_AL_51_2K_NULL_20210123_NUL_SMPTE_OV)
    . CPL XML syntax and structure check.
      Schema validation error : Element '{http://www.w3.org/2001/XMLSchema}import': Failed to parse the XML resource 'http://www.w3.org/TR/2002/REC-xmldsig-core-20020212/xmldsig-core-schema.xsd'., line 4
      Using schema : http://www.smpte-ra.org/schemas/429-7/2006/CPL
```

**Comment by bioleyl** [comment]
Here is the complete console print of my test with you last change.

I added time elapsed for each steps on the lines.

The time elapsed between the line "Checking DCP" and the start of the hash is 2minutes. 
The time elapsed between the end of the hash and the print of the report is 4minutes.

When I make a test on the linux machine, thoses times are barely 1-2 seconds not minutes,

I hope this will help. Thanks for your time on this.

```
PS C:\worker_corrupt\test lib\ClairMeta> python -m clairmeta.cli check -type dcp C:\testfiles\tstdcp\CAVALES_SHR_S_FR-XX-EN_AL_51_2K_NULL_20210123_NUL_SMPTE_OV -progress
2021-02-09 09:24:17,569 - Clairmeta - INFO - Probing DCP : C:\testfiles\tstdcp\CAVALES_SHR_S_FR-XX-EN_AL_51_2K_NULL_20210123_NUL_SMPTE_OV
2021-02-09 09:24:37,971 - Clairmeta - INFO - Total time : 20.40 seconds
2021-02-09 09:24:38,047 - Clairmeta - INFO - Checking DCP : C:\testfiles\tstdcp\CAVALES_SHR_S_FR-XX-EN_AL_51_2K_NULL_20210123_NUL_SMPTE_OV <-- This line is printed 21 sec after the job started
[  13.92 GiB in 12.66 sec (at 1179.88 MBytes/s)              ] 100.00% - 8f2eeddd-ac5b-483d-ba52-4b4a6bb3a353_video.mxf <-- Hash is starting 2:22 after the start of the job. 2 min elapsed before last line
[  985.24 MiB in 0.87 sec (at 1180.98 MBytes/s)              ] 100.00% - 8c3ddcb6-f489-4d22-a0a1-f50ebb0b5462_audio.mxf <-- Hash is finish at 2:35
2021-02-09 09:30:52,268 - Clairmeta - INFO - Check report: <-- Report is printed 6:35 after the start of the job. It took 4min between the end of the hash and the report printed

Status : Fail
Path : C:\testfiles\tstdcp\CAVALES_SHR_S_FR-XX-EN_AL_51_2K_NULL_20210123_NUL_SMPTE_OV
Size : 14.88 GiB
Total check : 122
Total time : 374.21 sec

Error(s):
  + PKL_7cb7d1ba-007e-4d31-8fcc-5ba6b214a62b.xml (CAVALES_SHR_S_FR-XX_AL_51_2K_NULL_20210123_NUL_SMPTE_OV)
    . PKL XML syntax and structure check.
      Schema validation error : Element '{http://www.w3.org/2001/XMLSchema}import': Failed to parse the XML resource 'http://www.w3.org/TR/2002/REC-xmldsig-core-20020212/xmldsig-core-schema.xsd'., line 4
      Using schema : http://www.smpte-ra.org/schemas/429-8/2007/PKL
    + 2a9d8f9e-8799-4e5e-a6c2-9c38967bb0c5_sub.mxf
      . PKL assets hash check.
        Check unknown error
        Traceback (most recent call last):
          File "C:\worker_corrupt\test lib\ClairMeta\clairmeta\utils\file.py", line 145, in __call__
            human_size(file_size), file_elapsed, (file_size / 1e6) / file_elapsed)
        ZeroDivisionError: float division by zero

        During handling of the above exception, another exception occurred:

        Traceback (most recent call last):
          File "C:\worker_corrupt\test lib\ClairMeta\clairmeta\dcp_check_base.py", line 152, in run_check
            check_res = check(*args)
          File "C:\worker_corrupt\test lib\ClairMeta\clairmeta\dcp_check_pkl.py", line 111, in check_assets_pkl_hash
            self.hash_map[asset_id] = shaone_b64(path, self.hash_callback)
          File "C:\worker_corrupt\test lib\ClairMeta\clairmeta\utils\file.py", line 199, in shaone_b64
            callback(file_path, run_size, file_size, time_cb - start)
          File "C:\worker_corrupt\test lib\ClairMeta\clairmeta\utils\file.py", line 147, in __call__
            raise ZeroDivisionError(
        ZeroDivisionError: ConsoleProgress callback error: float division by zero Size 0.369812 Elapsed 0.0

    + CPL_29bea761-1123-49e5-90ae-786c61e0d8cd.xml (CAVALES_SHR_S_FR-EN_AL_51_2K_NULL_20210123_NUL_SMPTE_OV)
      . PKL assets hash check.
        Check unknown error
        Traceback (most recent call last):
          File "C:\worker_corrupt\test lib\ClairMeta\clairmeta\utils\file.py", line 145, in __call__
            human_size(file_size), file_elapsed, (file_size / 1e6) / file_elapsed)
        ZeroDivisionError: float division by zero

        During handling of the above exception, another exception occurred:

        Traceback (most recent call last):
          File "C:\worker_corrupt\test lib\ClairMeta\clairmeta\dcp_check_base.py", line 152, in run_check
            check_res = check(*args)
          File "C:\worker_corrupt\test lib\ClairMeta\clairmeta\dcp_check_pkl.py", line 111, in check_assets_pkl_hash
            self.hash_map[asset_id] = shaone_b64(path, self.hash_callback)
          File "C:\worker_corrupt\test lib\ClairMeta\clairmeta\utils\file.py", line 199, in shaone_b64
            callback(file_path, run_size, file_size, time_cb - start)
          File "C:\worker_corrupt\test lib\ClairMeta\clairmeta\utils\file.py", line 147, in __call__
            raise ZeroDivisionError(
        ZeroDivisionError: ConsoleProgress callback error: float division by zero Size 0.011332 Elapsed 0.0

    + CPL_a60452a4-14e0-479d-b9f9-4f69e1f6fde7.xml (CAVALES_SHR_S_FR-XX_AL_51_2K_NULL_20210123_NUL_SMPTE_OV)
      . PKL assets hash check.
        Check unknown error
        Traceback (most recent call last):
          File "C:\worker_corrupt\test lib\ClairMeta\clairmeta\utils\file.py", line 145, in __call__
            human_size(file_size), file_elapsed, (file_size / 1e6) / file_elapsed)
        ZeroDivisionError: float division by zero

        During handling of the above exception, another exception occurred:

        Traceback (most recent call last):
          File "C:\worker_corrupt\test lib\ClairMeta\clairmeta\dcp_check_base.py", line 152, in run_check
            check_res = check(*args)
          File "C:\worker_corrupt\test lib\ClairMeta\clairmeta\dcp_check_pkl.py", line 111, in check_assets_pkl_hash
            self.hash_map[asset_id] = shaone_b64(path, self.hash_callback)
          File "C:\worker_corrupt\test lib\ClairMeta\clairmeta\utils\file.py", line 199, in shaone_b64
            callback(file_path, run_size, file_size, time_cb - start)
          File "C:\worker_corrupt\test lib\ClairMeta\clairmeta\utils\file.py", line 147, in __call__
            raise ZeroDivisionError(
        ZeroDivisionError: ConsoleProgress callback error: float division by zero Size 0.010908 Elapsed 0.0

  + CPL_29bea761-1123-49e5-90ae-786c61e0d8cd.xml (CAVALES_SHR_S_FR-EN_AL_51_2K_NULL_20210123_NUL_SMPTE_OV)
    . CPL XML syntax and structure check.
      Schema validation error : Element '{http://www.w3.org/2001/XMLSchema}import': Failed to parse the XML resource 'http://www.w3.org/TR/2002/REC-xmldsig-core-20020212/xmldsig-core-schema.xsd'., line 4
      Using schema : http://www.smpte-ra.org/schemas/429-7/2006/CPL
    + 2a9d8f9e-8799-4e5e-a6c2-9c38967bb0c5_sub.mxf
      . Subtitle font file exists.
        Subtitle missing font file : ad48b4af-9e02-4cb6-8867-b81ea5e1816d
  + CPL_a60452a4-14e0-479d-b9f9-4f69e1f6fde7.xml (CAVALES_SHR_S_FR-XX_AL_51_2K_NULL_20210123_NUL_SMPTE_OV)
    . CPL XML syntax and structure check.
      Schema validation error : Element '{http://www.w3.org/2001/XMLSchema}import': Failed to parse the XML resource 'http://www.w3.org/TR/2002/REC-xmldsig-core-20020212/xmldsig-core-schema.xsd'., line 4
      Using schema : http://www.smpte-ra.org/schemas/429-7/2006/CPL
Warning(s):
  + CPL_29bea761-1123-49e5-90ae-786c61e0d8cd.xml (CAVALES_SHR_S_FR-EN_AL_51_2K_NULL_20210123_NUL_SMPTE_OV)
    . CPL ContentTitleText and PKL AnnotationText shall match.
      Multi CPLs package shall use a common denominator of all CPL titles as the PKL AnnotationText : CAVALES_SHR_S_FR-EN_AL_51_2K_NULL_20210123_NUL_SMPTE_OV / CAVALES_SHR_S_FR-XX_AL_51_2K_NULL_20210123_NUL_SMPTE_OV

DCP - C:\testfiles\tstdcp\CAVALES_SHR_S_FR-XX-EN_AL_51_2K_NULL_20210123_NUL_SMPTE_OV - Check failed
```

**Comment by remia** [comment]
Hi @bioleyl,

Thanks for your help in debugging this!

I pushed yet another update that should avoid these errors.

I would be curious to see how that affect the time of the check and whether these 2 other errors are still here. It still something we need to address are these are not present in linux / macOS checks.

**Comment by remia** [comment]
Hey @bioleyl,

Did you get a chance to try that PR ?

Or are you just disabling progress report ?

**Comment by bioleyl** [comment]
Sorry for the delay of my answer.

I did try the PR and the 0 division problems are gone but the DCP fail even when it should not. 

We had to switch the server to linux because I think it was a problem with the Windows server not your software. The validation of the same DCP is a success on Linux.

When I got time, I will setup another Windows server for test but for now we had to continue the development.

Best regards and thanks for you time and reactivité. We will continue the tests on other DCP.
